silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/Auth0ProviderSpec.scala

```diff
@@ -0,0 +1,175 @@
+import play.api.libs.ws.{ WSRequest, WSResponse }
+import play.api.mvc.Results
+import play.api.test.{ FakeRequest, WithApplication }
+import play.mvc.Http
+import test.Helper
+
+import scala.concurrent.Future
+
+/**
+ * Test case for the [[Auth0Provider]] class.
+ */
+class Auth0ProviderSpec extends OAuth2ProviderSpec {
+
+  val AUTH0_ERROR_JSON: String = "providers/custom/auth0.error.json"
+  val AUTH0_SUCCESS_JSON: String = "providers/custom/auth0.success.json"
+  val AUTH0_USER_PROFILE_JSON: String = "providers/custom/auth0.profile.json"
+
+  val AUTH0_USER_INFO_URL = "https://mydomain.eu.auth0.com/userinfo"
+
+  "The `withSettings` method" should {
+    "create a new instance with customized settings" in new WithApplication with Context {
+      val s = provider.withSettings { s =>
+        s.copy(accessTokenURL = "new-access-token-url")
+      }
+
+      s.settings.accessTokenURL must be equalTo "new-access-token-url"
+    }
+  }
+
+  "The `authenticate` method" should {
+    "fail with UnexpectedResponseException if OAuth2Info can be build because of an unexpected response" in new WithApplication with Context {
+      val requestHolder = mock[WSRequest]
+      val response = mock[WSResponse]
+      implicit val req = FakeRequest(GET, "?" + Code + "=my.code")
+      response.json returns Json.obj()
+      requestHolder.withHeaders(any) returns requestHolder
+      requestHolder.post[Map[String, Seq[String]]](any)(any) returns Future.successful(response)
+      httpLayer.url(oAuthSettings.accessTokenURL) returns requestHolder
+      stateProvider.validate(any, any) returns Future.successful(state)
+
+      failed[UnexpectedResponseException](provider.authenticate()) {
+        case e => e.getMessage must startWith(InvalidInfoFormat.format(provider.id, ""))
+      }
+    }
+
+    "return the auth info" in new WithApplication with Context {
+      val requestHolder = mock[WSRequest]
+      val response = mock[WSResponse]
+      implicit val req = FakeRequest(GET, "?" + Code + "=my.code")
+      response.json returns oAuthInfo
+      requestHolder.withHeaders(any) returns requestHolder
+      requestHolder.post[Map[String, Seq[String]]](any)(any) returns Future.successful(response)
+      httpLayer.url(oAuthSettings.accessTokenURL) returns requestHolder
+      stateProvider.validate(any, any) returns Future.successful(state)
+
+      authInfo(provider.authenticate()) {
+        case authInfo => authInfo must be equalTo oAuthInfo.as[OAuth2Info]
+      }
+    }
+  }
+
+  "The `retrieveProfile` method" should {
+    "fail with ProfileRetrievalException if API returns error" in new WithApplication with Context {
+      val requestHolder = mock[WSRequest]
+      val response = mock[WSResponse]
+      private val statusCode: Int = Http.Status.UNAUTHORIZED
+      response.status returns statusCode
+      requestHolder.get() returns Future.successful(response)
+      val oauth2Info = oAuthInfo.as[OAuth2Info]
```

Comment:
Could you please define the variables above the mock expectation definitions. This would make the code more readable, it's divided into logical blocks. In the other tests too.